### PR TITLE
Change data value encoding to Base64 URL-safe

### PR DESCRIPTION
**Pull request by adamsero**

### CODE REVIEW

**Comment by adamsero** [comment]
PR for #1325

**Comment by sbernard31** [comment]
1. some concern about decoding, we started to discuss about it at : https://github.com/eclipse/leshan/issues/1325#issuecomment-1282437770

2. I think we should add tests about this (integration or unit test)

3. About issue reference in commit message, until recently I was using `#1325` but it is sometime painful because if you start the commit header with # sometime it can be consider as a commented line (see https://stackoverflow.com/questions/2788092/start-a-git-commit-message-with-a-hashmark#2788642) So recently I decide to use `GH-1325` instead of `#1325` in commit message. (see https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/autolinked-references-and-urls#issues-and-pull-requests).
If you think this is a good idea, please replace `eclipse#1325` by `GH-1325`.

File: leshan-core/src/main/java/org/eclipse/leshan/core/util/base64/DefaultBase64Encoder.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Override
+    public String encode(byte[] dataToEncode) {
+        Base64.Encoder encoder = urlSafe ? Base64.getUrlEncoder() : Base64.getEncoder();
+        if (withoutPadding) {
+            encoder = encoder.withoutPadding();
+        }
+
+        return encoder.encodeToString(dataToEncode);
+    }
```
Comment (sbernard31) [inline leshan-core/src/main/java/org/eclipse/leshan/core/util/base64/DefaultBase64Encoder.java:37]: maybe better to create the encoder internally at constructor time ?
(If you agree this must be also change for Decoder)

File: leshan-core/src/main/java/org/eclipse/leshan/senml/json/jackson/SenMLJsonRecordSerDes.java [hunk 2]
```diff
@@ -105,7 +106,7 @@ else if (value instanceof Float) {
                 jsonObj.put("vlo", record.getObjectLinkValue());
                 break;
             case OPAQUE:
-                jsonObj.put("vd", Base64.encodeBase64String(record.getOpaqueValue()));
+                jsonObj.put("vd", new DefaultBase64Encoder(true, true).encode(record.getOpaqueValue()));
```
Comment (sbernard31) [inline leshan-core/src/main/java/org/eclipse/leshan/senml/json/jackson/SenMLJsonRecordSerDes.java:109]: Should we store the encoder in an attribute ?
(same for decoder ?)

File: leshan-core/src/main/java/org/eclipse/leshan/senml/json/jackson/SenMLJsonRecordSerDes.java [hunk 3]
```diff
@@ -171,7 +172,7 @@ public SenMLRecord deserialize(JsonNode o) throws JsonException {
 
         JsonNode vd = o.get("vd");
         if (vd != null && vd.isTextual()) {
-            record.setOpaqueValue(Base64.decodeBase64(vd.asText()));
+            record.setOpaqueValue(new DefaultBase64Decoder(true, true).decode(vd.asText()));
```
Comment (sbernard31) [inline leshan-core/src/main/java/org/eclipse/leshan/senml/json/jackson/SenMLJsonRecordSerDes.java:175]: Should we catch the `IllegalArgumentException` and raise a `JsonException` instead ?

File: leshan-core/src/main/java/org/eclipse/leshan/senml/json/jackson/SenMLJsonJacksonEncoderDecoder.java [hunk 4]
```diff
@@ -75,7 +75,7 @@ public SenMLPack fromSenML(byte[] jsonString) throws SenMLException {
             }
             return new SenMLPack(serDes.deserialize(node.iterator()));
 
-        } catch (JsonException | IOException e) {
+        } catch (JsonException | IllegalArgumentException | IOException e) {
```
Comment (sbernard31) [inline leshan-core/src/main/java/org/eclipse/leshan/senml/json/jackson/SenMLJsonJacksonEncoderDecoder.java:78]: (If we encapsulate `IllegalArgumentException` in  `JsonException`   we should revert this change)

**Review by sbernard31 — COMMENTED** [review]
I think we should add some test about base64 decoder /encoder.

Maybe with case raised at : https://github.com/eclipse/leshan/pull/676#issuecomment-481188756

File: leshan-core/src/main/java/org/eclipse/leshan/senml/json/jackson/SenMLJsonRecordSerDes.java (hunk 2 above)
Comment (adamsero) [reply leshan-core/src/main/java/org/eclipse/leshan/senml/json/jackson/SenMLJsonRecordSerDes.java:109]: We actually discussed it in #1348 in [this comment](https://github.com/eclipse/leshan/issues/1348#issue-1447767894):

>And finally in our custom version of SenMLJsonRecordSerDes, we basically have 2 lines of code to change with respect to the Leshan version (related to OPAQUE fields encoding and decoding):

```java
case OPAQUE:
    jsonObj.put("vd", urlSafeEncoder.encode(record.getOpaqueValue()));
    break;
```
>and
```java
JsonNode vd = o.get("vd");
if (vd != null && vd.isTextual()) {
    record.setOpaqueValue(urlSafeDecoder.decode(vd.asText()));
    hasValue = true;
}
```

So that would require better customization of `SenMLJsonRecordSerDes` class and others.

Comment (sbernard31) [reply leshan-core/src/main/java/org/eclipse/leshan/senml/json/jackson/SenMLJsonRecordSerDes.java:109]: I'm not sure I get you :thinking: 

My point was just to avoid a decoder(or encoder) object creation by value to encode or decode.

I'm not sure this is related to better customization ? 

Or maybe you mean : "I will change this later anyway so we don't need to clean this now ?"

**Comment by sbernard31** [comment]
You didn't answer to : 

> I think we should add some test about base64 decoder /encoder.
> Maybe with case raised at : https://github.com/eclipse/leshan/pull/676#issuecomment-481188756

So I  don't know if you plan to do it or not ?  :slightly_smiling_face:

File: leshan-core/src/main/java/org/eclipse/leshan/core/util/base64/DefaultBase64Encoder.java [hunk 5]
```diff
@@ -12,27 +12,28 @@
...
  *     Orange Polska S.A. - initial API and implementation
  *******************************************************************************/
-
 package org.eclipse.leshan.core.util.base64;
 
 import java.util.Base64;
 
 public class DefaultBase64Encoder implements Base64Encoder {
     private final boolean urlSafe;
     private final boolean withoutPadding;
```
Comment (sbernard31) [inline leshan-core/src/main/java/org/eclipse/leshan/core/util/base64/DefaultBase64Encoder.java:21]: attributes seems useless now ?

File: leshan-core/src/main/java/org/eclipse/leshan/core/util/base64/DefaultBase64Decoder.java [hunk 6]
```diff
@@ -19,10 +19,12 @@
 public class DefaultBase64Decoder implements Base64Decoder {
     private final boolean urlSafe;
     private final boolean withoutPadding;
+    private final Base64.Decoder decoder;
 
     public DefaultBase64Decoder(boolean urlSafe, boolean withoutPadding) {
         this.urlSafe = urlSafe;
```
Comment (sbernard31) [inline leshan-core/src/main/java/org/eclipse/leshan/core/util/base64/DefaultBase64Decoder.java:25]: urlSafe attribute seems useless now.

File: leshan-core/src/main/java/org/eclipse/leshan/core/util/base64/DefaultBase64Decoder.java [hunk 7]
```diff
@@ -39,8 +41,6 @@ public String decode(byte[] encoded) {
         if (!this.withoutPadding) {
```
Comment (sbernard31) [inline leshan-core/src/main/java/org/eclipse/leshan/core/util/base64/DefaultBase64Decoder.java:41]: you still create a new decoder on decode.

**Review by sbernard31 — CHANGES_REQUESTED** [review]

**Comment by adamsero** [comment]
@sbernard31 Just want to let you know that @Warmek will finish this pull request as well as any future tasks as I'm leaving Orange and the project. Thank you for your time spent on helping me learn the code and the project overall 🙂.

**Comment by sbernard31** [comment]
@adamsero, never a good news to see contributor go. I hope this is mainly a desired choice from you.
Thanks for all your contributions and good luck for the future. :pray: 

If you have any feedback about how we worked together on the project (good or bad things) please do not hesitate to share. This could help to improve.

**Comment by sbernard31** [comment]
@Warmek do no hesitate to ask if you face any issue / difficulty on this :wink:

File: leshan-core/src/main/java/org/eclipse/leshan/core/util/base64/DefaultBase64Decoder.java (hunk 7 above)
Comment (Warmek) [reply leshan-core/src/main/java/org/eclipse/leshan/core/util/base64/DefaultBase64Decoder.java:38]: I don't understand where do i create new decoder on decode?

Comment (sbernard31) [reply leshan-core/src/main/java/org/eclipse/leshan/core/util/base64/DefaultBase64Decoder.java:38]: Comment review seems to not be attached to the right line, so this doesn't help to understand :grimacing: 

In this class, there is 2 `decode` methods.
1 method reuses decoder created in constructor, the other recreate a new decoder on each `decode` call.

Decoder should be create only once in constructor for both methods.

Tell me if this is not clearer enough.

Comment (Warmek) [reply leshan-core/src/main/java/org/eclipse/leshan/core/util/base64/DefaultBase64Decoder.java:38]: As far as i can tell the only new instances are two new strings that are created in this method. I can't see where in this method could new decoder be created 🤔
```java
    @Override
    public String decode(byte[] encoded) {
        if (!this.withoutPadding) {
            validateEncodedData(new String(encoded));
        }
        byte[] decoded = decoder.decode(encoded);
        return new String(decoded);
    }
```

Comment (sbernard31) [reply leshan-core/src/main/java/org/eclipse/leshan/core/util/base64/DefaultBase64Decoder.java:38]: This is in the other method. 
```diff
public byte[] decode(String encoded) {
        if (!this.withoutPadding) {
            validateEncodedData(encoded);
        }
--        Base64.Decoder decoder = urlSafe ? Base64.getUrlDecoder() : Base64.getDecoder();
        return decoder.decode(encoded);
}
```
(note that I was thinking that  `Base64.getUrlDecoder()` and` Base64.getDecoder() ` created a new decoder but this is a static reference. Anyway, I think it's more consistent to reuse the attribute create in constructor.

Maybe could even be replaced by : 
```diff
public byte[] decode(String encoded) {
--        if (!this.withoutPadding) {
--            validateEncodedData(encoded);
--        }
--       Base64.Decoder decoder = urlSafe ? Base64.getUrlDecoder() : Base64.getDecoder();
--        return decoder.decode(encoded);
++   return decode(encoded.getByte());
}
```

Not directly linked but I just see that Base64.decode(String) consider that String is encoded in ISO_8859_1 which is very strange to me. I will try to understand why...

**Comment by sbernard31** [comment]
@Warmek let's me know when this is in a reviewable state :)

**Comment by Warmek** [comment]
@sbernard31 I think i've done it and it's ready for your review

File: leshan-core/src/main/java/org/eclipse/leshan/senml/json/jackson/SenMLJsonRecordSerDes.java (hunk 2 above)
Comment (Warmek) [reply leshan-core/src/main/java/org/eclipse/leshan/senml/json/jackson/SenMLJsonRecordSerDes.java:109]: I think it depends if we will call decoders/encoders with different options often or not. I probably would lean towards:
```java
jsonObj.put("vd", Base64.encodeBase64String(record.getOpaqueValue()));
```
But if we aim to add more options for  decoders/encoders or change those options often, we should probably stick to creating new decoders/encoders.

Comment (sbernard31) [reply leshan-core/src/main/java/org/eclipse/leshan/senml/json/jackson/SenMLJsonRecordSerDes.java:109]: We want to be able to let user creates a `SenMLJsonRecordSerDes` with a defirent base64 decoder than the one by default. (discussion at #1348)

But I don't think we want to be able to change `SenMLJsonRecordSerDes` behavior dynamically.

I mean once we create a `SenMLJsonRecordSerDes` with a given base64 decoder we don't change it.


So my idea was to create an attribute initialized with `new DefaultBase64Encoder(true, true)` then use it here.

Comment (Warmek) [reply leshan-core/src/main/java/org/eclipse/leshan/senml/json/jackson/SenMLJsonRecordSerDes.java:109]: In that case i would initialize  decoder once and then reuse it

**Review by sbernard31 — APPROVED** [review]

**Comment by sbernard31** [comment]
@Warmek, I will take some time to get deeper about the charset encoding question. (just to be sure I well understand the problem)

Then I plan to work a bit more on this : 
- removing `org.eclipse.leshan.core.util.Base64` class
- maybe add more test and validation in base64 codec.
- maybe add a decoder which is able to support all format for LWM2M old-json  Content Format. 

I let you know when I have code to review.

**Review by Warmek — APPROVED** [review]
👍

**Comment by Warmek** [comment]
@sbernard31 Do you think it will be possibile to share leshan M10 in first half of december?

**Comment by sbernard31** [comment]
> Do you think it will be possibile to share leshan M10 in first half of december?

I'm not sure what means "share" ? 
If you talk about a release, we can try but this means we have just 1 week more left.

And after my proposition at : https://github.com/eclipse/leshan/issues/1319#issuecomment-1271378134
You (at Orange) asked in an email for more time to test before the release : 
> We as Orange would prefer to know a week ahead before a milestone is released so we can test all the features for possible bugs.

You (Orange) also proposes : 
>  Maybe we could create a separate issue for that purpose or organize it differently if you'd like.

So I created an issue talk about 2.0.0-M10 release (as I did for 2.0.0-M9) : #1370

**Review by Warmek — APPROVED** [review]
👍

**Comment by Warmek** [comment]
@sbernard31 I reviewed your changes and i wait for merge

**Comment by sbernard31** [comment]
This is now integrated in #1371, so I close this PR.

Thx @adamsero and @Warmek  for your contributions :pray: !

**Comment by JaroslawLegierski** [comment]
> You (at Orange) asked in an email for more time to test before the release :
> 
> > We as Orange would prefer to know a week ahead before a milestone is released so we can test all the features for possible bugs.

Yes our colleagues are ready to test all new features - as soon as they are available in the master branch.

> Thx @adamsero and @Warmek for your contributions 🙏 !

First of all, we thank You for Your help !

**Comment by sbernard31** [comment]
> First of all, we thank You for Your help !

Glad to help :slightly_smiling_face:

**Comment by sbernard31** [comment]
@Warmek  @JaroslawLegierski 

I fall on this last week : https://www.eclipse.org/projects/handbook/#resources-commit

> The email address of the author must match the email address on the Eclipse Foundation Account, the name "Some Body" must be replaced by the real name of the person.

:point_up:  this is not checked by any eclipse validation tools  but would be better if you change your git configuration to match the eclipse requirement. 

Currently @Warmek  `Author` looks like  `Author: Warmek <b***********@orange.com>` (which is not so good)
and @JaroslawLegierski  looks like this : `Author: JaroslawLegierski <j*********@orange.com> ` (which is almost OK but maybe better with a space)

I let you know but I think this is clearly not critical.

**Comment by JaroslawLegierski** [comment]
> and @JaroslawLegierski looks like this : `Author: JaroslawLegierski <j*********@orange.com> ` (which is almost OK but maybe better with a space)

I modified the Name field in the Github configuration -> now is in the form FirstName space LastName. I'll have to see what is the effect of these changes.

**Comment by Warmek** [comment]
I've changed my user.name via command: 
git config --global user.name "Some Body" 

Will that help?

**Comment by sbernard31** [comment]
did you commit your last commit (d0861f804693083e9503733b7366aa379569cd81) before or after running this ?
(because in your last commit there is still `Author: Warmek <b***********@orange.com>`

**Comment by Warmek** [comment]
I've made my last commit before running this command

**Comment by sbernard31** [comment]
OK when you will push a new commit, I will be able to say you if this is OK now.  :slightly_smiling_face:
